Index: linux-4.1.52/net/netfilter/nf_conntrack_core.c
===================================================================
--- linux-4.1.52.orig/net/netfilter/nf_conntrack_core.c	2019-03-15 16:57:40.066999276 +0800
+++ linux-4.1.52/net/netfilter/nf_conntrack_core.c	2019-03-25 17:49:00.731585777 +0800
@@ -964,7 +964,7 @@
 			}
 		}
 
-		ila = kzalloc(sizeof(struct nf_sess_ref_count),GFP_KERNEL);
+		ila = kzalloc(sizeof(struct nf_sess_ref_count),GFP_ATOMIC);
 		if(!ila){
 			printk(KERN_ERR "Unable to create nf_sess_ref_count \n");
 			return -ENOMEM;
