Index: libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_general.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/wireless/zcfg_fe_dal_wifi_general.c	2018-09-07 13:35:26.600754300 +0800
+++ libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_general.c	2018-09-07 13:40:51.985117404 +0800
@@ -454,15 +454,22 @@
 
 	if(wifiApObj) zcfgFeObjJsonSetWithoutApply(RDM_OID_WIFI_ACCESS_POINT, &wifiApIid, wifiApObj, NULL);
 	if(wifiSecObj) zcfgFeObjJsonSetWithoutApply(RDM_OID_WIFI_ACCESS_POINT_SEC, &wifiSecIid, wifiSecObj, NULL);
-#ifdef WIFI_ONESSID_SECURITY_SYNC
-	if(syncWifiSecObj) zcfgFeObjJsonBlockedSet(RDM_OID_WIFI_ACCESS_POINT_SEC, &syncWifiSecIid, syncWifiSecObj, NULL);
-#endif //WIFI_ONESSID_SECURITY_SYNC
 	if(wifiRadioObj) zcfgFeObjJsonSetWithoutApply(RDM_OID_WIFI_RADIO, &wifiRadioIid, wifiRadioObj, NULL);
 	if(wifiSsidObj)zcfgFeObjJsonBlockedSet(RDM_OID_WIFI_SSID, &wifiSsidIid, wifiSsidObj, NULL);
 	if(wifiSubnetObj) zcfgFeObjJsonBlockedSet(RDM_OID_DHCPV4_SRV_SUBNET, &wifiSubnetIid, wifiSubnetObj, NULL);
 #ifdef ZYXEL_ONESSID
 	if(wifiOneSsidObj) zcfgFeObjJsonBlockedSet(RDM_OID_WIFI_ONE_SSID, &wifiOneSsidIid, wifiOneSsidObj, NULL);
 #endif
+
+#ifdef WIFI_ONESSID_SECURITY_SYNC
+	if ((wifiOneSsidObj) && (json_object_get_boolean(json_object_object_get(wifiOneSsidObj, "Enable")) == true)) {
+		//do not thing >> sync sec skip
+	}
+	else{ //no suppoert ZYXEL_ONESSID or ONESSID is disable  >> sync sec active
+		if(syncWifiSecObj) zcfgFeObjJsonBlockedSet(RDM_OID_WIFI_ACCESS_POINT_SEC, &syncWifiSecIid, syncWifiSecObj, NULL);
+	}
+#endif //WIFI_ONESSID_SECURITY_SYNC
+
 	return ret;
 }
 
