Index: libzcfg_fe_dal-1.0/security/package3rdParty/zcfg_fe_dal_package3rdParty.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/security/package3rdParty/zcfg_fe_dal_package3rdParty.c	2018-12-27 19:06:17.700266944 +0800
+++ libzcfg_fe_dal-1.0/security/package3rdParty/zcfg_fe_dal_package3rdParty.c	2018-12-27 19:07:19.950564723 +0800
@@ -20,6 +20,7 @@
 	{"Enable",			dalType_boolean,	0,	0,	NULL,	NULL,	0},
 	{"Data", 			dalType_string, 	0,	0,	NULL,	NULL,	0},
 	{"FixCustID",		dalType_string, 	0,	0,	NULL,	NULL,	0},
+	{"Hash",		    dalType_string, 	0,	0,	NULL,	NULL,	0},
 	{"Status",			dalType_string, 	0,	0,	NULL,	NULL,	0},
 	{NULL,				0,					0,	0,	NULL,	NULL,	0}
 };
@@ -38,6 +39,7 @@
 	struct json_object *script = NULL;
 	struct json_object *data = NULL;
 	struct json_object *custID = NULL;
+	struct json_object *hash = NULL;
 #if 1 //Avast Agent
 	char buffer[1024] = {0};
 	char *scriptName = NULL;
@@ -57,6 +59,7 @@
 		enable = json_object_object_get(Package3rdPartyObj, "Enable");
 		data = json_object_object_get(Package3rdPartyObj, "Data");
 		custID = json_object_object_get(Package3rdPartyObj, "FixCustID");
+		hash = json_object_object_get(Package3rdPartyObj, "Hash");
 #if 1 //Avast Agnet
 		scriptName = json_object_get_string(json_object_object_get(Package3rdPartyObj, "Script"));
 		//printf("zcfgFeDalPackage3rdPartyGet : scriptName = %s\n", scriptName);
@@ -81,6 +84,7 @@
 		json_object_object_add(pramJobj, "Enable", JSON_OBJ_COPY(enable));
 		json_object_object_add(pramJobj, "Data", JSON_OBJ_COPY(data));
 		json_object_object_add(pramJobj, "FixCustID", JSON_OBJ_COPY(custID));
+		json_object_object_add(pramJobj, "Hash", JSON_OBJ_COPY(hash));
 #if 1 //Avast Agnet		
 		json_object_object_add(pramJobj, "Status", json_object_new_string(buffer));
 #else
