Index: libzcfg_fe_dal-1.0/voip/sip/zcfg_fe_dal_sip_account.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/voip/sip/zcfg_fe_dal_sip_account.c	2019-09-19 18:44:12.637417940 +0800
+++ libzcfg_fe_dal-1.0/voip/sip/zcfg_fe_dal_sip_account.c	2019-09-19 18:47:11.235994713 +0800
@@ -83,6 +83,9 @@
 const char* PhyReferenceList;
 const char* AuthUserName;
 const char* AuthPassword;
+#ifdef WIND_ITALY_CUSTOMIZATION
+const char hiddenAuthPassword[] = "********" ;
+#endif
 const char* X_ZYXEL_Url_Type;
 bool CallTransferEnable;
 bool CallWaitingEnable;
@@ -1171,7 +1174,20 @@
 		if(json_object_object_get(Jobj, "AuthUserName"))
 		json_object_object_add(lineSipObj, "AuthUserName", json_object_new_string(AuthUserName));
 		if(json_object_object_get(Jobj, "AuthPassword"))
+			
+#ifdef WIND_ITALY_CUSTOMIZATION
+		//update sip password
+		{
+			if (!strcmp(AuthPassword, hiddenAuthPassword)){ //it's mean no changes password by user
+				//no change password no save
+			}
+			else{
+				json_object_object_add(lineSipObj, "AuthPassword", json_object_new_string(AuthPassword));
+			}
+		}
+#else
 		json_object_object_add(lineSipObj, "AuthPassword", json_object_new_string(AuthPassword));
+#endif
 		if(json_object_object_get(Jobj, "Url_Type"))
 		json_object_object_add(lineSipObj, "X_ZYXEL_Url_Type", json_object_new_string(X_ZYXEL_Url_Type));
 
@@ -1788,7 +1804,19 @@
 		X_ZYXEL_Url_Type = json_object_get_string(json_object_object_get(lineSipObj, "X_ZYXEL_Url_Type"));
 
 		json_object_object_add(paserobj, "AuthUserName", json_object_new_string(AuthUserName));
+		
+#ifdef WIND_ITALY_CUSTOMIZATION
+		//Change hidden password to show in GUI >> 8 digit
+		//if not emty than show '********'
+		if (strlen(AuthPassword) > 0){
+			json_object_object_add(paserobj, "AuthPassword", json_object_new_string(hiddenAuthPassword));
+		}
+		else{ //orig
+			json_object_object_add(paserobj, "AuthPassword", json_object_new_string(AuthPassword));
+		}
+#else
 		json_object_object_add(paserobj, "AuthPassword", json_object_new_string(AuthPassword));
+#endif
 		json_object_object_add(paserobj, "X_ZYXEL_Url_Type", json_object_new_string(X_ZYXEL_Url_Type));
 
 		zcfgFeJsonObjFree(lineSipObj);
