Index: libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_general.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/wireless/zcfg_fe_dal_wifi_general.c	2019-09-19 18:43:50.044032056 +0800
+++ libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_general.c	2019-09-19 18:43:50.064032056 +0800
@@ -1022,7 +1022,9 @@
 				X_ZYXEL_AutoGenPSK = json_object_get_boolean(json_object_object_get(wifiSecObj, "X_ZYXEL_AutoGenPSK"));
 				psk_value = json_object_get_string(json_object_object_get(wifiSecObj, "X_ZYXEL_Auto_PSK"));
 			}
-			
+			/*Beacause SAEPassphrae just support in wpa3sae and wpa2pskwpa3sae,let SAEPassphrae="" in wpa2psk and wpawpa2psk*/
+			json_object_object_add(wifiSecObj, "SAEPassphrase", json_object_new_string(""));
+
 			if(X_ZYXEL_AutoGenPSK)
 			{
 				X_ZYXEL_Auto_PSK = json_object_get_string(json_object_object_get(wifiSecObj, "X_ZYXEL_Auto_PSK"));
@@ -1089,6 +1091,7 @@
 			{
 				X_ZYXEL_Auto_PSK = json_object_get_string(json_object_object_get(wifiSecObj, "X_ZYXEL_Auto_PSK"));
 				json_object_object_add(wifiSecObj, "KeyPassphrase", json_object_new_string(X_ZYXEL_Auto_PSK));
+				json_object_object_add(wifiSecObj, "SAEPassphrase", json_object_new_string(X_ZYXEL_Auto_PSK));
 				json_object_object_add(wifiSecObj, "PreSharedKey", json_object_new_string(""));
 			}
 			else {
@@ -1097,10 +1100,12 @@
 					memset(psk_lowercase, 0, strlen(psk_value)+1);
 					stolower(psk_value, psk_lowercase, strlen(psk_value));
 					json_object_object_add(wifiSecObj, "KeyPassphrase", json_object_new_string(""));
+					json_object_object_add(wifiSecObj, "SAEPassphrase", json_object_new_string(""));
 					json_object_object_add(wifiSecObj, "PreSharedKey", json_object_new_string(psk_lowercase));
 				}
 				else {
 					json_object_object_add(wifiSecObj, "KeyPassphrase", json_object_new_string(psk_value));
+					json_object_object_add(wifiSecObj, "SAEPassphrase", json_object_new_string(psk_value));
 				}
 			}
 		}
