Index: libzcfg_fe_dal-1.0/voip/sip/zcfg_fe_dal_sip_account.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/voip/sip/zcfg_fe_dal_sip_account.c	2019-12-12 08:42:36.878965902 +0000
+++ libzcfg_fe_dal-1.0/voip/sip/zcfg_fe_dal_sip_account.c	2019-12-12 08:45:08.309303865 +0000
@@ -1543,10 +1543,14 @@
 		if(replyMsg)
 		strcat(replyMsg, "Add function doesn't need Index.");
 	}
-	multiJobj = zcfgFeJsonObjNew();
+
 	if(prepareMultiJobj != NULL){
 		multiJobj = prepareMultiJobj;
 	}
+	else
+	{
+		multiJobj = zcfgFeJsonObjNew();
+	}
 
 	if((ret = zcfgFeDalLineObjAdd(Jobj)) != ZCFG_SUCCESS){
 		printf("%s:(ERROR) fail to add line obj \n", __FUNCTION__);
