diff -urN libzcfg_fe_dal-1.0.org/maintenance/useraccount/zcfg_fe_dal_user_account.c libzcfg_fe_dal-1.0/maintenance/useraccount/zcfg_fe_dal_user_account.c
--- libzcfg_fe_dal-1.0.org/maintenance/useraccount/zcfg_fe_dal_user_account.c	2019-12-25 16:55:25.895409800 +0800
+++ libzcfg_fe_dal-1.0/maintenance/useraccount/zcfg_fe_dal_user_account.c	2019-12-27 15:21:00.482960855 +0800
@@ -339,52 +339,86 @@
 		}
 
 		if(ISSET("Password")){
-			if(!strcmp("", Jgets(Jobj ,"Password"))){
-				curPassword = Jgets(accountObj, "Password");
-				if(strcmp(curPassword, Jgets(Jobj ,"oldPw"))!= 0){
-					if(replyMsg) strcpy(replyMsg, "Wrong --oldPw value");
-					json_object_object_add(Jobj, "__multi_lang_replyMsg", json_object_new_string("zylang.Common.oldpassword_err2"));
-						zcfgFeJsonObjFree(accountObj);
-					return ZCFG_NOT_FOUND; //GUI use ZCFG_NOT_FOUND to detect oldPw error.
-				}
-				replaceParam(accountObj, "Password", Jobj, "oldPw");
-			}
-			else{
-				if(!ISSET("oldPw") || !ISSET("verifyNewpassword")){
-					if(replyMsg) strcpy(replyMsg, "Need --oldPw and --verifyNewpassword to modify Password");
-					zcfgFeJsonObjFree(accountObj);
-					return ZCFG_DELETE_REJECT;
-				}
-
-				if(strcmp(Jgets(Jobj, "Password"), Jgets(Jobj, "verifyNewpassword"))){
-					if(replyMsg) strcpy(replyMsg, "--Password and --verifyNewpassword don't match");
-					json_object_object_add(Jobj, "__multi_lang_replyMsg", json_object_new_string("zylang.Common.notMatch"));
-					zcfgFeJsonObjFree(accountObj);
-					return ZCFG_REQUEST_REJECT;
-				}
-				
-				authResult = zyUtilAuthUserAccount(username, Jgets(Jobj ,"oldPw"));
-				if (authResult == 1) 
-				{
-					curPassword = Jgets(accountObj, "Password");
-					if(strcmp(curPassword, Jgets(Jobj ,"oldPw"))){
-						if(replyMsg) strcpy(replyMsg, "Wrong --oldPw value");
-						json_object_object_add(Jobj, "__multi_lang_replyMsg", json_object_new_string("zylang.Common.oldpassword_err2"));
+			if(strcmp("", Jgets(Jobj ,"Password"))){//if psw is not empty
+				if(ISSET("oldPw")){
+					if(!strcmp("", Jgets(Jobj ,"oldPw"))){//if old psw is empty
+						if(replyMsg) {
+							strcpy(replyMsg, "Need --oldPw to modify Password");
+						}
+						json_object_object_add(Jobj, "__multi_lang_replyMsg", 
+							json_object_new_string("zylang.Common.oldPassword_empty"));
 						zcfgFeJsonObjFree(accountObj);
 						return ZCFG_NOT_FOUND; //GUI use ZCFG_NOT_FOUND to detect oldPw error.
 					}
+					else{
+						if(ISSET("verifyNewpassword")){//if verify psw is empty
+							if(!strcmp("", Jgets(Jobj ,"verifyNewpassword"))){
+								if(replyMsg) {
+									strcpy(replyMsg, "Need --verifyNewpassword to modify Password");
+								}
+								json_object_object_add(Jobj, "__multi_lang_replyMsg", 
+									json_object_new_string("zylang.Common.verify_password_empty"));
+								zcfgFeJsonObjFree(accountObj);
+								return ZCFG_REQUEST_REJECT;
+							}
+							else{
+								//check if verify psw is correct
+								if(strcmp(Jgets(Jobj, "Password"), Jgets(Jobj, "verifyNewpassword"))){
+									if(replyMsg) 
+										strcpy(replyMsg, "--Password and --verifyNewpassword don't match");
+									json_object_object_add(Jobj, "__multi_lang_replyMsg", 
+										json_object_new_string("zylang.Common.notMatch"));
+									zcfgFeJsonObjFree(accountObj);
+									return ZCFG_REQUEST_REJECT;
+								}
+
+								//use username to check current psw
+								authResult = zyUtilAuthUserAccount(username, Jgets(Jobj ,"oldPw"));
+								if(authResult == 1){
+									curPassword = Jgets(accountObj, "Password");
+									if(strcmp(curPassword, Jgets(Jobj ,"oldPw"))!= 0){
+										if(replyMsg) 
+											strcpy(replyMsg, "Wrong --oldPw value");
+										json_object_object_add(Jobj, "__multi_lang_replyMsg", 
+											json_object_new_string("zylang.Common.oldpassword_err2"));
+										zcfgFeJsonObjFree(accountObj);
+										return ZCFG_NOT_FOUND; //GUI use ZCFG_NOT_FOUND to detect oldPw error.
+									}
+								}
+								else if (authResult < 0){
+									if(replyMsg) 
+										strcpy(replyMsg, "Wrong --oldPw value");
+									json_object_object_add(Jobj, "__multi_lang_replyMsg", 
+										json_object_new_string("zylang.Common.oldpassword_err2"));
+									zcfgFeJsonObjFree(accountObj);
+									return ZCFG_NOT_FOUND; //GUI use ZCFG_NOT_FOUND to detect oldPw error.
+								}							
+								//if current psw and verify psw is checked, update data model psw
+								replaceParam(accountObj, "Password", Jobj, "Password");
+							}
+						}
+						else{
+							if(replyMsg) {
+								strcpy(replyMsg, "Need --verifyNewpassword to modify Password");
+							}
+							json_object_object_add(Jobj, "__multi_lang_replyMsg", 
+								json_object_new_string("zylang.Common.verify_password_empty"));
+							zcfgFeJsonObjFree(accountObj);
+							return ZCFG_REQUEST_REJECT;
+						}
+					}
 				}
-				else if (authResult < 0)
-				{
-					if(replyMsg) strcpy(replyMsg, "Wrong --oldPw value");
-					json_object_object_add(Jobj, "__multi_lang_replyMsg", json_object_new_string("zylang.Common.oldpassword_err2"));
+				else{
+					if(replyMsg) {
+						strcpy(replyMsg, "Need --oldPw to modify Password");
+					}
+					json_object_object_add(Jobj, "__multi_lang_replyMsg", 
+						json_object_new_string("zylang.Common.oldPassword_empty"));
 					zcfgFeJsonObjFree(accountObj);
 					return ZCFG_NOT_FOUND; //GUI use ZCFG_NOT_FOUND to detect oldPw error.
 				}
-				replaceParam(accountObj, "Password", Jobj, "Password");
 			}
 		}
-		
 		replaceParam(accountObj, "Enabled", Jobj, "Enabled");
 		replaceParam(accountObj, "AccountRetryTime", Jobj, "AccountRetryTime");
 		replaceParam(accountObj, "AccountIdleTime", Jobj, "AccountIdleTime");
