Index: libzcfg_fe_dal-1.0/zcfg_fe_dal.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/zcfg_fe_dal.c	2019-11-19 16:26:39.257628916 +0800
+++ libzcfg_fe_dal-1.0/zcfg_fe_dal.c	2019-11-19 16:39:36.777551819 +0800
@@ -140,6 +140,10 @@
 extern dal_param_t SFP_MGMT_param[];
 #endif
 
+#ifdef ZYXEL_WEB_GUI_SHOW_ZYEE /* JoannaSu, 20191115, support ZyEE configuration */
+extern dal_param_t ZYEE_DU_param[];
+#endif
+
 extern zcfgRet_t zcfgFeDalWan(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
 extern zcfgRet_t zcfgFeDalMgmt(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
 extern zcfgRet_t zcfgFeDalDnsEntry(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
@@ -267,6 +271,9 @@
 #ifdef ZYXEL_SFP_POWER_MGMT
 extern zcfgRet_t zcfgFeDalSfpMgmt(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
 #endif
+#ifdef ZYXEL_WEB_GUI_SHOW_ZYEE /* JoannaSu, 20191115, support ZyEE configuration */
+extern zcfgRet_t zcfgFeDalDUAction(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg);
+#endif
 
 
 
@@ -497,6 +504,9 @@
 #ifdef ZYXEL_SFP_POWER_MGMT
 {"sfp_mgmt", SFP_MGMT_param, zcfgFeDalSfpMgmt, 	zcfgFeDalShowSfpMgmt,		"edit|get",				"SFPManagement"},
 #endif
+#ifdef ZYXEL_WEB_GUI_SHOW_ZYEE /* JoannaSu, 20191115, support ZyEE configuration */
+{"ZYDUACT", ZYEE_DU_param, zcfgFeDalDUAction,  NULL,       "edit",             ""},
+#endif
 {NULL,				NULL,				NULL, 					NULL,						NULL,				 NULL}
 };
 
Index: libzcfg_fe_dal-1.0/Makefile
===================================================================
--- libzcfg_fe_dal-1.0.orig/Makefile	2019-11-19 16:26:39.257628916 +0800
+++ libzcfg_fe_dal-1.0/Makefile	2019-11-19 21:54:50.050830492 +0800
@@ -1,7 +1,7 @@
 CURRENT_DIR = $(shell pwd)
 COMMON_ZCFG_INCLUDE_DIR = $(CURRENT_DIR)/../../../private/apps/zcfg/include
 LIB_PATH = -L$(USRLIBDIR) -L$(INSTALL_DIR)/lib/public -L$(INSTALL_DIR)/lib/private
-LINK_LIB = -ljson -lzcfg_msg -lzcfg_fe_rdm_access -lzcfg_fe_rdm_string -lzcfg_fe_rdm_struct -lzcfg_fe_schema -lzyutil -lzlog
+LINK_LIB = -ljson -lzcfg_msg -lzcfg_fe_rdm_access -lzcfg_fe_rdm_string -lzcfg_fe_rdm_struct -lzcfg_fe_schema -lzyutil -lzlog -lzos
 INC_PATH = -I$(TOOLCHAIN)/include -I$(COMMON_ZCFG_INCLUDE_DIR) -I.
 
 FE_PATH = ./
@@ -24,6 +24,7 @@
 MAINTENANCE_EMAIL_PATH = ./maintenance/mail
 MAINTENANCE_LOG_PATH = ./maintenance/logsetting
 MAINTENANCE_USER_ACCOUNT_PATH = ./maintenance/useraccount
+MAINTENANCE_ZYEE_DU_PATH = ./maintenance/zyee
 MAINTENANCE = ./maintenance
 SECURITY_PPTP_PATH = ./security/pptp
 VOIP_PATH = ./voip
@@ -65,6 +66,9 @@
 SOURCES += $(wildcard $(MAINTENANCE_MGMT)/*.c)
 SOURCES += $(wildcard $(MAINTENANCE_DIAGNOSTIC_PATH)/*.c)
 SOURCES += $(wildcard $(MAINTENANCE_USER_ACCOUNT_PATH)/*.c)
+ifeq ($(CONFIG_ZYXEL_WEB_GUI_SHOW_ZYEE),y)
+SOURCES += $(wildcard $(MAINTENANCE_ZYEE_DU_PATH)/*.c)
+endif
 SOURCES += $(wildcard $(MAINTENANCE)/*.c)
 ifeq ($(CONFIG_KERNEL_ZYXEL_PPTP_SUPPORT),y)
 SOURCES += $(wildcard $(SECURITY_PPTP_PATH)/*.c)
@@ -375,6 +379,11 @@
 CFLAGS += -DZYXEL_SFP_POWER_MGMT
 endif
 
+#ZyEE
+ifeq ($(CONFIG_ZYXEL_WEB_GUI_SHOW_ZYEE),y)
+CFLAGS += -DZYXEL_WEB_GUI_SHOW_ZYEE
+endif
+
 ifeq ($(CONFIG_PACKAGE_mos),y)
 CFLAGS += -DMOS_SUPPORT
 endif
@@ -397,7 +406,6 @@
 
 .c.o:
 	$(CC) $(CFLAGS) -Wall -fPIC -c $< -o $@
-
 libzcfg_fe_dal : $(OBJS)
 	$(CC) $(LDFLAGS) -shared -o libzcfg_fe_dal.so $(OBJS)
 
Index: libzcfg_fe_dal-1.0/maintenance/zyee/zcfg_fe_dal_zyeeduaction.c
===================================================================
--- /dev/null	1970-01-01 00:00:00.000000000 +0000
+++ libzcfg_fe_dal-1.0/maintenance/zyee/zcfg_fe_dal_zyeeduaction.c	2019-11-19 16:30:01.729991757 +0800
@@ -0,0 +1,219 @@
+#include <json/json.h>
+
+#include "zcfg_common.h"
+#include "zcfg_fe_rdm_access.h"
+#include "zcfg_fe_rdm_string.h"
+#include "zcfg_fe_rdm_struct.h"
+#include "zcfg_rdm_oid.h"
+#include "zcfg_rdm_obj.h"
+#include "zcfg_debug.h"
+#include "zcfg_msg.h"
+#include "zcfg_fe_dal_common.h"
+#include "zlog_api.h"
+#include "zcfg_system.h"
+#include "zos_api.h"
+
+dal_param_t ZYEE_DU_param[]={
+    {"OPType",                   dalType_string,    0,  0,  NULL},
+    {"UUID",                     dalType_string, 0,  0,  NULL},
+    {"URL",                      dalType_string, 0,  0,  NULL},
+    {"Username",                 dalType_string, 0,  0,  NULL},
+    {"Password",                 dalType_int,    0,  0,  NULL},
+    {"SelectedEE",                dalType_string, 0,  0,  NULL},
+    {"DUName",                dalType_string, 0,  0,  NULL},
+    {NULL,                       0,              0,  0,  NULL},
+};
+
+#define ZYEE_DU_INSTALL_RST_PATH    "/tmp/gui_zyee_du_install_.log"
+#define ZYEE_DU_UNINSTALL_RST_PATH  "/tmp/gui_zyee_du_uninstall.log"
+
+zcfgRet_t zcfgFeDalDUActionGet(struct json_object *Jobj, struct json_object *Jarray, char *replyMsg)
+{    zcfgRet_t       ret = ZCFG_SUCCESS;
+     FILE            *fp = NULL;
+     char            rcvbuf[256] = {0};
+     char            *rcvptr = NULL;
+     int             rstIndex = 0;
+     char            strStatus[640] = {0};
+
+     const char      *opType;
+     struct json_object *paramJobj = NULL;
+     paramJobj = json_object_new_object();
+
+     if ( NULL != json_object_object_get(Jobj, "OPType") )
+     {
+         opType = json_object_get_string(json_object_object_get(Jobj, "OPType"));
+     }
+
+     if (strcmp(opType,"Uninstall") == 0)
+     {
+         fp = fopen(ZYEE_DU_UNINSTALL_RST_PATH, "r");
+     }
+     else if (strcmp(opType,"Install") == 0)
+     {
+         fp = fopen(ZYEE_DU_INSTALL_RST_PATH, "r");
+     }
+     else
+     {
+         return ZCFG_INTERNAL_ERROR;
+     }
+
+     if (fp == NULL)
+     {
+         return ZCFG_INTERNAL_ERROR;
+     }
+
+     memset(strStatus, 0, sizeof(strStatus));
+     /* get the response */
+    while ((fgets(rcvbuf, 256, fp)) != NULL)
+    {
+        ZLOG_INFO("Recive: %s", rcvbuf);
+        if (rcvptr = strstr(rcvbuf, "zyeecmd: "))
+        {
+            rstIndex++;
+            strcat(strStatus,rcvptr);
+            continue;
+        }
+        if(rstIndex)
+        {
+            strcat(strStatus,rcvbuf);
+            break;
+        }
+        memset(rcvbuf, 0, sizeof(rcvbuf));
+    }
+
+    if (strlen(strStatus) > 0)
+    {
+        json_object_object_add(paramJobj, "OPResult", json_object_new_string(strStatus));
+        json_object_object_add(paramJobj, "OPState", json_object_new_string("Complete"));
+    }
+    else
+    {
+        if (strcmp(opType,"Uninstall") == 0)
+        {
+            json_object_object_add(paramJobj, "OPState", json_object_new_string("Uninstalling"));
+        }
+        else
+        {
+            json_object_object_add(paramJobj, "OPState", json_object_new_string("Installing"));
+        }
+
+        json_object_object_add(paramJobj, "OPResult", json_object_new_string(""));
+    }
+    json_object_array_add(Jarray, paramJobj);
+
+    return ret;
+}
+
+zcfgRet_t zcfgFeDalDUActionSet(struct json_object *Jobj, char *replyMsg)
+{
+    zcfgRet_t       ret = ZCFG_SUCCESS;
+    char            eidStr[4] = {0};
+    char            cmd[1792] = {0};
+
+    const char      *opType;
+    const char      *uuid;
+    const char      *url;
+    const char      *username;
+    const char      *password;
+    const char      *eeName;
+    const char      *duName;
+
+    ZLOG_INFO("input --- %s \n ",json_object_to_json_string(Jobj));
+
+    if ( NULL != json_object_object_get(Jobj, "OPType") )
+    {
+        opType = json_object_get_string(json_object_object_get(Jobj, "OPType"));
+    }
+    if ( NULL != json_object_object_get(Jobj, "UUID") )
+    {
+        uuid = json_object_get_string(json_object_object_get(Jobj, "UUID"));
+    }
+    if( NULL != json_object_object_get(Jobj, "URL") )
+    {
+        url = json_object_get_string(json_object_object_get(Jobj, "URL"));
+    }
+    if ( NULL != json_object_object_get(Jobj, "Username") )
+    {
+        username = json_object_get_string(json_object_object_get(Jobj, "Username"));
+    }
+    if ( NULL != json_object_object_get(Jobj, "Password") )
+    {
+        password = json_object_get_string(json_object_object_get(Jobj, "Password"));
+    }
+    if ( NULL != json_object_object_get(Jobj, "NotifyEEName") )
+    {
+        eeName = json_object_get_string(json_object_object_get(Jobj, "NotifyEEName"));
+    }
+
+    if ( NULL != json_object_object_get(Jobj, "DUName") )
+    {
+        duName = json_object_get_string(json_object_object_get(Jobj, "DUName"));
+    }
+
+    if(strcmp(opType,"Uninstall") == 0)
+    {
+        sprintf(cmd, "zyeecmd notify %s -m ", eeName);
+        strcat(cmd, "\"uninstall ");
+        strcat(cmd, duName);
+        strcat(cmd, "\" -eid ");
+        sprintf(eidStr,"%d", ZCFG_EID_ZHTTPD);
+        strcat(cmd, eidStr);
+        strcat(cmd, " > ");
+        strcat(cmd, ZYEE_DU_UNINSTALL_RST_PATH);
+        strcat(cmd, " &");
+    }
+    else if(strcmp(opType,"Install") == 0)
+    {
+        sprintf(cmd, "zyeecmd notify %s -m ", eeName);
+        strcat(cmd, "\"install ");
+        strcat(cmd, url);
+        strcat(cmd, " -uuid ");
+        strcat(cmd, uuid);
+        if (strlen(username) != 0)
+        {
+            strcat(cmd, " -u ");
+            strcat(cmd, username);
+            strcat(cmd, ":");
+            strcat(cmd, password);
+        }
+
+        strcat(cmd, "\" -eid ");
+        sprintf(eidStr,"%d", ZCFG_EID_ZHTTPD);
+        strcat(cmd, eidStr);
+        strcat(cmd, " > ");
+        strcat(cmd, ZYEE_DU_INSTALL_RST_PATH);
+        strcat(cmd, " &");
+    }
+    ZLOG_INFO("cmd --- %s \n ",cmd);
+
+    if(ZOS_SYSTEM(cmd) == false)
+    {
+        return ZCFG_INTERNAL_ERROR;
+    }
+
+    return ret;
+}
+
+zcfgRet_t zcfgFeDalDUAction(const char *method, struct json_object *Jobj, struct json_object *Jarray, char *replyMsg)
+{
+    zcfgRet_t ret = ZCFG_SUCCESS;
+
+    if (!method || !Jobj)
+    {
+        return ZCFG_INTERNAL_ERROR;
+    }
+
+    if (!strcmp(method, "PUT"))
+    {
+        ret = zcfgFeDalDUActionSet(Jobj, NULL);
+    }
+    else if (!strcmp(method, "GET"))
+    {
+        ret = zcfgFeDalDUActionGet(Jobj, Jarray, NULL);
+    }
+
+
+
+
+    return ret;
+}
