Index: libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_general.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/wireless/zcfg_fe_dal_wifi_general.c	2020-06-04 16:49:17.284833885 +0800
+++ libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_general.c	2020-06-04 17:26:02.476982100 +0800
@@ -2170,8 +2170,6 @@
 	bool autoGenMoreSecurePasswd;
 	bool wpsEnabled = false;
 	bool isMainSSID;
-	int preset_backhaul_count = 0;
-	int TOTAL_INTERFACE_5G_ssid;
 
 	ssidIndex = json_object_get_int(json_object_object_get(Jobj,"Index"));
 	IID_INIT(ssidIid);
@@ -2245,19 +2243,6 @@
 			return ZCFG_SUCCESS;
 	}
 
-	// get ssid obj
-	IID_INIT(ssidIid);
-	ssidIid.level = 1;
-	for(ssidIid.idx[0] = 1; ssidIid.idx[0] <= (TOTAL_INTERFACE_24G + TOTAL_INTERFACE_5G); ssidIid.idx[0]++){
-		if(zcfgFeObjJsonGetWithoutUpdate(RDM_OID_WIFI_SSID, &ssidIid, &ssidObj) == ZCFG_SUCCESS) {
-			if(json_object_get_int(json_object_object_get(ssidObj, "X_ZYXEL_Attribute"))) {
-				preset_backhaul_count += 1;
-			}
-		}
-		zcfgFeJsonObjFree(ssidObj);
-	}
-	TOTAL_INTERFACE_5G_ssid = TOTAL_INTERFACE_5G - preset_backhaul_count;
-
 	// get parameters
 	Index = json_object_get_int(json_object_object_get(Jobj, "Index"));
 	isMainSSID = json_object_get_boolean(json_object_object_get(wifiSsidObj, "X_ZYXEL_MainSSID"));
@@ -2267,7 +2252,7 @@
 #ifdef ZYXEL_SUPPORT_8021X
 	if(Index >= 1 && Index <= TOTAL_INTERFACE_24G)
 		wifiWPSIid.idx[0] = 1;
-	else if(Index >= (TOTAL_INTERFACE_24G + 1) && Index <= (TOTAL_INTERFACE_24G + TOTAL_INTERFACE_5G_ssid))
+	else if(Index >= (TOTAL_INTERFACE_24G + 1) && Index <= (TOTAL_INTERFACE_24G + TOTAL_INTERFACE_5G))
 	{
 		wifiWPSIid.idx[0] = (TOTAL_INTERFACE_24G + 1);
 	}
Index: libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_wmm.c
===================================================================
--- libzcfg_fe_dal-1.0.orig/network/wireless/zcfg_fe_dal_wifi_wmm.c	2020-06-04 16:49:17.200833886 +0800
+++ libzcfg_fe_dal-1.0/network/wireless/zcfg_fe_dal_wifi_wmm.c	2020-06-04 17:27:14.251982100 +0800
@@ -55,8 +55,8 @@
 	struct json_object *wirelessMode5G = NULL;
 	struct json_object *wifiRadioObj = NULL;
 	struct json_object *ssidObj = NULL;
-	int preset_backhaul_count = 0;
-	int TOTAL_INTERFACE_5G_ssid;
+	const char *SSIDReference = NULL;
+	bool hiddenIntf = false;
 
 	IID_INIT(iid);
 	while(zcfgFeObjJsonGetNextWithoutUpdate(RDM_OID_WIFI_RADIO, &iid, &wifiRadioObj) == ZCFG_SUCCESS){
@@ -79,19 +79,6 @@
 #endif
 	}
 
-	// get ssid obj
-	IID_INIT(ssidIid);
-	ssidIid.level = 1;
-	for(ssidIid.idx[0] = 1; ssidIid.idx[0] <= (TOTAL_INTERFACE_24G + TOTAL_INTERFACE_5G); ssidIid.idx[0]++){
-		if(zcfgFeObjJsonGetWithoutUpdate(RDM_OID_WIFI_SSID, &ssidIid, &ssidObj) == ZCFG_SUCCESS) {
-			if(json_object_get_int(json_object_object_get(ssidObj, "X_ZYXEL_Attribute"))) {
-				preset_backhaul_count += 1;
-			}
-		}
-		zcfgFeJsonObjFree(ssidObj);
-	}
-	TOTAL_INTERFACE_5G_ssid = TOTAL_INTERFACE_5G - preset_backhaul_count;
-
 	IID_INIT(iid);
 	int count = 1;
 	while(zcfgFeObjJsonGetNextWithoutUpdate(RDM_OID_WIFI_ACCESS_POINT, &iid, &obj) == ZCFG_SUCCESS){
@@ -114,14 +101,20 @@
 			json_object_object_add(paramJobj, "X_ZYXEL_Wireless_Mode", json_object_new_string(""));
 		}
 #else
-		if(iid.idx[0] <= TOTAL_INTERFACE_24G){
+		IID_INIT(ssidIid);
+		ssidIid.level = 1;
+		if((SSIDReference = json_object_get_string(json_object_object_get(obj, "SSIDReference"))) != NULL
+			&& sscanf(SSIDReference, "WiFi.SSID.%hhu", &(ssidIid.idx[0])) == 1
+			&& zcfgFeObjJsonGetWithoutUpdate(RDM_OID_WIFI_SSID, &ssidIid, &ssidObj) == ZCFG_SUCCESS) {
+			if(json_object_get_int(json_object_object_get(ssidObj, "X_ZYXEL_Attribute")))
+				hiddenIntf = true;
+			zcfgFeJsonObjFree(ssidObj);
+		}
+
+		if(!hiddenIntf && iid.idx[0] <= TOTAL_INTERFACE_24G){
 			json_object_object_add(paramJobj, "X_ZYXEL_Wireless_Mode", json_object_new_string(json_object_get_string(wirelessMode24G)));
 		}
-#if defined(BUILD_SONIQ) || defined(ZYXEL_EASYMESH)
-		else if(iid.idx[0] >= (TOTAL_INTERFACE_24G + 1) && iid.idx[0] <= (TOTAL_INTERFACE_24G + TOTAL_INTERFACE_5G - 2)) //-2 :exclude MultyPro & EasyMesh
-#else
-		else if(iid.idx[0] >= (TOTAL_INTERFACE_24G + 1) && iid.idx[0] <= (TOTAL_INTERFACE_24G + TOTAL_INTERFACE_5G))
-#endif
+		else if(!hiddenIntf && iid.idx[0] >= (TOTAL_INTERFACE_24G + 1) && iid.idx[0] <= (TOTAL_INTERFACE_24G + TOTAL_INTERFACE_5G))
 		{
 			json_object_object_add(paramJobj, "X_ZYXEL_Wireless_Mode", json_object_new_string(json_object_get_string(wirelessMode5G)));
 		}else{
@@ -129,6 +122,7 @@
 		}
 #endif
 		json_object_array_add(Jarray, paramJobj);
+		hiddenIntf = false;
 		count++;
 		zcfgFeJsonObjFree(obj);
 	}
