Index: libzcfg_fe_tr98-1.0/ispService.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/ispService.c	2018-12-07 18:31:39.082589108 +0800
+++ libzcfg_fe_tr98-1.0/ispService.c	2018-12-07 18:32:40.098718601 +0800
@@ -11,10 +11,6 @@
 #include "zcfg_debug.h"
 #include "zcfg_fe_tr98.h"
 #include "ispService_parameter.h"
-#if 1 //Avast Agnet
-#define SYSCMD "sh /misc/etc/init.d/avast-agent.sh status"
-#endif
-
 
 extern tr98Object_t tr98Obj[];
 
@@ -31,6 +27,8 @@
 	tr98Parameter_t *paramList = NULL;
 #if 1 //Avast Agent
 	char buffer[1024] = {0};
+	char sysCmd[512] = {0};
+	char *scriptName = NULL;
 	FILE *fp = NULL;
 #endif
 	zcfgLog(ZCFG_LOG_DEBUG, "Enter %s\n", __FUNCTION__);
@@ -50,8 +48,12 @@
 		paramValue = json_object_object_get(ispSrvObj, paramList->name);
 		if(paramValue != NULL) {	
 #if 1 //Avast Agent
+			if (strcmp(paramList->name, "Script") == 0){
+				scriptName = json_object_get_string(paramValue);
+				sprintf(sysCmd, "sh /misc/etc/init.d/%s status", scriptName);
+			}
 			if (strcmp(paramList->name, "Status") == 0){
-				if((fp = popen(SYSCMD, "r")) != NULL && fgets(buffer, sizeof(buffer), fp) != NULL) {
+				if((fp = popen(sysCmd, "r")) != NULL && fgets(buffer, sizeof(buffer), fp) != NULL) {
 					printf("zcfgFeDalPackage3rdPartyGet : syscmd get status buffer: %s\n", buffer);
 				}
 				json_object_object_add(*tr98Jobj, paramList->name, json_object_new_string(buffer));
