Index: libzcfg_fe_tr98-1.0/ispService.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/ispService.c	2018-12-27 17:03:22.948063749 +0800
+++ libzcfg_fe_tr98-1.0/ispService.c	2018-12-27 17:09:23.599310709 +0800
@@ -88,7 +88,9 @@
 	struct json_object *tmpObj = NULL;
 	tr98Parameter_t *paramList = NULL;
 	bool bEnable = false;
-
+	#if 1 //WIND request : enable agent and "Hash" can not be NULL
+	char *pHash = NULL;
+	#endif
 	zcfgLog(ZCFG_LOG_DEBUG, "Enter %s\n", __FUNCTION__);
 
 	IID_INIT(ispSrvIid);
@@ -113,6 +115,11 @@
 			if (strcmp(paramList->name, "Enable") == 0){
 				bEnable = json_object_get_boolean(paramValue);
 			}
+			#if 1 //WIND request : enable agent and "Hash" can not be NULL
+			if (strcmp(paramList->name, "Hash") == 0){
+				pHash = json_object_get_string(paramValue);
+			}
+			#endif
 			if(tr181ParamValue != NULL) {
 				json_object_object_add(ispSrvObj, paramList->name, JSON_OBJ_COPY(paramValue));
 				if (strcmp(paramList->name, "Enable") == 0){
@@ -129,6 +136,13 @@
 		SetParenCtlDisableForce();
 	}
 
+	#if 1 //WIND request : enable agent and "Hash" can not be NULL
+	if ((strlen(pHash) == 0) && (bEnable == true)){
+		printf("%s : Hash is NULL >> %s = %d  \n", __FUNCTION__, paramList->name, bEnable);
+		json_object_object_add(ispSrvObj, "Enable", json_object_new_boolean(false));
+	}
+	#endif
+
 	/*Set */
 	if(multiJobj){
 		json_object_put(tmpObj);
