Index: libzcfg_fe_tr98-1.0/Makefile
===================================================================
--- libzcfg_fe_tr98-1.0.orig/Makefile	2019-12-04 18:59:37.558366311 +0800
+++ libzcfg_fe_tr98-1.0/Makefile	2019-12-04 19:00:07.739841491 +0800
@@ -149,6 +149,11 @@
 CFLAGS += -DZYPRJ_CUSTOMER_WIND=1
 endif
 
+# Bandwidth Improvement
+ifeq ($(CONFIG_ZYXEL_BANDWIDTH_IMPROVEMENT),y)
+CFLAGS += -DZYXEL_BANDWIDTH_IMPROVEMENT=1
+endif
+
 #ISKON
 ifeq ($(CONFIG_USE_OLD_VOIP_PREFIX),y)
 CFLAGS += -DUSE_OLD_VOIP_PREFIX
Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98.c	2019-12-04 17:36:46.654229348 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98.c	2019-12-04 17:39:41.167468057 +0800
@@ -364,6 +364,10 @@
 {e_TR98_PROXIMITY_SENSOR		       , &tr98Obj[e_TR98_PROXIMITY_SENSOR	     -e_TR98_START]},
 #endif //#ifdef PROXIMITY_SENSOR_SUPPORT
 #endif
+#ifdef ZYXEL_BANDWIDTH_IMPROVEMENT
+{e_TR98_BANDWIDTH_IMPROVEMENT , &tr98Obj[e_TR98_BANDWIDTH_IMPROVEMENT	     -e_TR98_START]},
+#endif //ZYXEL_BANDWIDTH_IMPROVEMENT
+
 #if OBSOLETED
 #else
 //{e_TR98_WAN_DSL_CONN_MGMT       , &tr98Obj[e_TR98_WAN_DSL_CONN_MGMT       -e_TR98_START]},
Index: libzcfg_fe_tr98-1.0/zcfg_fe_tr98_handle.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zcfg_fe_tr98_handle.h	2019-12-04 17:36:46.654229348 +0800
+++ libzcfg_fe_tr98-1.0/zcfg_fe_tr98_handle.h	2019-12-04 17:39:55.438790555 +0800
@@ -279,6 +279,10 @@
 #define TR98_PROXIMITY_SENSOR "InternetGatewayDevice.X_ZYXEL_EXT.ProximitySensor"
 #endif //#ifdef PROXIMITY_SENSOR_SUPPORT
 
+#ifdef ZYXEL_BANDWIDTH_IMPROVEMENT
+#define TR98_BANDWIDTH_IMPROVEMENT "InternetGatewayDevice.X_ZYXEL_EXT.BandwidthImprovement"
+#endif //ZYXEL_BANDWIDTH_IMPROVEMENT
+
 #ifdef BUILD_SONIQ
 #define TR98_Qsteer			"InternetGatewayDevice.X_ZYXEL_EXT.Qsteer"
 #endif
@@ -679,6 +683,9 @@
 #ifdef PROXIMITY_SENSOR_SUPPORT
 {TR98_PROXIMITY_SENSOR,	0, 	para_ProximitySensor,	zyExtProximitySensorObjGet,		zyExtProximitySensorObjSet,		   NULL,             NULL,            NULL},
 #endif //#ifdef PROXIMITY_SENSOR_SUPPORT
+#ifdef ZYXEL_BANDWIDTH_IMPROVEMENT
+{TR98_BANDWIDTH_IMPROVEMENT,	0, 	para_BandwidthImprovement,	zyExtBandwidthImprovementObjGet,		zyExtBandwidthImprovementObjSet,		   NULL,             NULL,            NULL},
+#endif //ZYXEL_BANDWIDTH_IMPROVEMENT
 #endif
 #if OBSOLETED
 #else
Index: libzcfg_fe_tr98-1.0/zyExt.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zyExt.c	2019-12-04 17:36:46.566229346 +0800
+++ libzcfg_fe_tr98-1.0/zyExt.c	2019-12-04 17:40:03.655445652 +0800
@@ -7205,6 +7205,132 @@
 #endif //#ifdef PROXIMITY_SENSOR_SUPPORT
 
 
+#ifdef ZYXEL_BANDWIDTH_IMPROVEMENT
+/* InternetGatewayDevice.X_ZYXEL_EXT.BandwidthImprovement */
+zcfgRet_t zyExtBandwidthImprovementObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag)
+{
+	zcfgRet_t ret;
+	objIndex_t objIid;
+	struct json_object *improvementObj = NULL;
+	struct json_object *paramValue = NULL;
+	tr98Parameter_t *paramList = NULL;
+	char *status = NULL;
+
+	printf("%s : Enter\n", __FUNCTION__);
+	
+	enum {
+		Active = 0,
+		OFF,
+		OFF_NoDSLConnect,
+		OFF_NoMobileConnect,
+		OFF_DataRate128k
+	};
+	
+	/*Get related tr181 objects*/
+	IID_INIT(objIid);
+	if((ret = feObjJsonGet(RDM_OID_ZY_BAND_WIDTH_IMPROVEMENT, &objIid, &improvementObj, updateFlag)) != ZCFG_SUCCESS){
+		return ret;
+	}
+
+	/*fill up tr98 object from related tr181 objects*/
+	*tr98Jobj = json_object_new_object();
+	paramList = tr98Obj[handler].parameter;
+	while(paramList->name != NULL){
+		/*Write parameter value from tr181 objects to tr98 object*/		
+		paramValue = json_object_object_get(improvementObj, paramList->name);
+		if(paramValue != NULL){
+			if (!strcmp(paramList->name, "Status")){
+				if (json_object_get_int(paramValue) == Active){
+					strcpy(status, "Active");
+				}
+				else if (json_object_get_int(paramValue) == OFF_NoDSLConnect){
+					strcpy(status, "OFF: No DSL Connection");
+				}
+				else if (json_object_get_int(paramValue) == OFF_NoMobileConnect){
+					strcpy(status, "OFF: No Mobile Device");
+				}
+				else {
+					strcpy(status, "OFF");
+				}
+				printf("### zyExtBandwidthImprovementObjGet ### GET status = %s\n");
+				json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(status));
+				
+			} 
+			else{
+				json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(paramValue));
+			}
+			paramList++;
+			continue;
+		}
+
+		/*Not defined in tr181, give it a default value*/
+		printf("Can't find parameter %s in TR181\n", paramList->name);
+		paramList++;	
+	}
+	
+	json_object_put(improvementObj);
+	
+	return ZCFG_SUCCESS;
+}
+
+zcfgRet_t zyExtBandwidthImprovementObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj, char *paramfault)
+{
+	zcfgRet_t ret;
+	objIndex_t objIid;
+	struct json_object *improvementObj = NULL;
+	struct json_object *tmpObj = NULL;
+	struct json_object *paramValue = NULL;
+	struct json_object *tr181ParamValue = NULL;
+	tr98Parameter_t *paramList = NULL;
+
+	printf("%s : Enter\n", __FUNCTION__);
+
+	/*Get related tr181 objects*/
+	IID_INIT(objIid);
+	if((ret = zcfgFeObjJsonGet(RDM_OID_ZY_BAND_WIDTH_IMPROVEMENT, &objIid, &improvementObj)) != ZCFG_SUCCESS){
+		return ret;
+	}
+
+	if(multiJobj){
+		tmpObj = improvementObj;
+		improvementObj = NULL;
+		improvementObj = zcfgFeJsonMultiObjAppend(RDM_OID_ZY_BAND_WIDTH_IMPROVEMENT, &objIid, multiJobj, tmpObj);
+	}
+
+	paramList = tr98Obj[handler].parameter;
+	while(paramList->name != NULL){
+		/*Write new parameter value from tr98 object to tr181 objects*/	
+		paramValue = json_object_object_get(tr98Jobj, paramList->name);
+		if(paramValue != NULL){
+			tr181ParamValue = json_object_object_get(improvementObj, paramList->name);
+			if(tr181ParamValue != NULL){
+				json_object_object_add(improvementObj, paramList->name, JSON_OBJ_COPY(paramValue));
+				paramList++;
+				continue;
+			}
+		}
+		printf("Can't find parameter %s in TR181\n", paramList->name);
+
+		paramList++;	
+	}
+
+	if(multiJobj){
+		json_object_put(tmpObj);
+	}
+	else{
+		if((ret = zcfgFeObjJsonSet(RDM_OID_ZY_BAND_WIDTH_IMPROVEMENT, &objIid, improvementObj, NULL)) != ZCFG_SUCCESS ){
+			json_object_put(improvementObj);
+			return ret;
+		}
+		json_object_put(improvementObj);
+	}
+	
+	return ZCFG_SUCCESS;
+}
+
+#endif
+
+
 /*
  *   TR98 Object Name : InternetGatewayDevice.X_ZYXEL_EXT.PrintServer
  *
Index: libzcfg_fe_tr98-1.0/zyExt.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zyExt.h	2019-12-04 17:36:46.566229346 +0800
+++ libzcfg_fe_tr98-1.0/zyExt.h	2019-12-04 17:40:13.791361063 +0800
@@ -57,6 +57,9 @@
 extern tr98Parameter_t para_ProximitySensor[];
 #endif //#ifdef PROXIMITY_SENSOR_SUPPORT
 
+#ifdef ZYXEL_BANDWIDTH_IMPROVEMENT
+extern tr98Parameter_t para_BandwidthImprovement[];
+#endif //ZYXEL_BANDWIDTH_IMPROVEMENT
 
 extern tr98Parameter_t para_Landing_Page[];
 extern tr98Parameter_t para_Feature_Flag[];
@@ -321,6 +324,10 @@
 extern zcfgRet_t zyExtProximitySensorObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj);
 #endif //#ifdef PROXIMITY_SENSOR_SUPPORT
 
+#ifdef ZYXEL_BANDWIDTH_IMPROVEMENT
+extern zcfgRet_t zyExtBandwidthImprovementObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag);
+extern zcfgRet_t zyExtBandwidthImprovementObjSet(char *tr98FullPathName, int handler, struct json_object *tr98Jobj, struct json_object *multiJobj);
+#endif //ZYXEL_BANDWIDTH_IMPROVEMENT
 
 #ifdef ZYXEL_LANDING_PAGE_FEATURE
 extern zcfgRet_t zyExtLandingPageObjGet(char *tr98FullPathName, int handler, struct json_object **tr98Jobj, bool updateFlag);
Index: libzcfg_fe_tr98-1.0/zyExt_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zyExt_parameter.h	2019-12-04 17:36:46.578229346 +0800
+++ libzcfg_fe_tr98-1.0/zyExt_parameter.h	2019-12-04 17:40:22.334929126 +0800
@@ -233,6 +233,17 @@
 };
 #endif //#ifdef PROXIMITY_SENSOR_SUPPORT
 
+#ifdef ZYXEL_BANDWIDTH_IMPROVEMENT
+tr98Parameter_t para_BandwidthImprovement[] = {
+	{ "Enable", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+	{ "Status", PARAMETER_ATTR_READONLY, 0, json_type_uint32},
+	{ "BWMInterval", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
+	{ "TriggerLording", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
+	{ "DeTriggerLording", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
+	{ NULL, 0, 0, 0}
+};
+#endif //ZYXEL_BANDWIDTH_IMPROVEMENT
+
 tr98Parameter_t para_SysInfo[] = {
 	{ "HostName", PARAMETER_ATTR_WRITE, 65, json_type_string},
 	{ "IsHostNameModify", PARAMETER_ATTR_WRITE, 0, json_type_boolean},