Index: libzcfg_fe_tr98-1.0/zyExt.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/zyExt.c	2019-12-05 13:35:06.080210381 +0800
+++ libzcfg_fe_tr98-1.0/zyExt.c	2019-12-05 13:36:33.539060569 +0800
@@ -7235,11 +7235,18 @@
 	/*fill up tr98 object from related tr181 objects*/
 	*tr98Jobj = json_object_new_object();
 	paramList = tr98Obj[handler].parameter;
+	#if 0 //zyExtBandwidthImprovementObjGet
+	printf("%s : Enter - 2 \n", __FUNCTION__);
+	#endif
 	while(paramList->name != NULL){
 		/*Write parameter value from tr181 objects to tr98 object*/		
 		paramValue = json_object_object_get(improvementObj, paramList->name);
+		#if 0 //zyExtBandwidthImprovementObjGet
+		printf("%s : Enter - 2 : paramList->name = %s\n", __FUNCTION__, paramList->name);
+		#endif
 		if(paramValue != NULL){
-			if (!strcmp(paramList->name, "Status")){
+			//if (!strcmp(paramList->name, "Status")){
+			if (0){ //skip 
 				if (json_object_get_int(paramValue) == Active){
 					strcpy(status, "Active");
 				}
@@ -7252,7 +7259,7 @@
 				else {
 					strcpy(status, "OFF");
 				}
-				printf("### zyExtBandwidthImprovementObjGet ### GET status = %s\n");
+				//printf("### zyExtBandwidthImprovementObjGet ### GET status = %s\n", status);
 				json_object_object_add(*tr98Jobj, paramList->name, JSON_OBJ_COPY(status));
 				
 			} 
