Index: libzcfg_fe_tr98-1.0/lanDev_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev_parameter.h	2019-09-10 11:22:12.252148107 +0800
+++ libzcfg_fe_tr98-1.0/lanDev_parameter.h	2019-09-10 11:25:28.545794602 +0800
@@ -320,6 +320,7 @@
 	{ "X_ZYXEL_ReauthInterval", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 	{ "X_ZYXEL_RekeyingInterval", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 #endif
+	{ "X_ZYXEL_InActiveChannels", PARAMETER_ATTR_WRITE, 129, json_type_string},
 	{ NULL, 0, 0, 0}
 };
 
