Index: libzcfg_fe_tr98-1.0/lanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev.c	2019-10-02 13:57:25.315332155 +0800
+++ libzcfg_fe_tr98-1.0/lanDev.c	2019-10-02 14:41:23.139332155 +0800
@@ -67,7 +67,7 @@
 #ifdef WIFISECURITY_AUTH_WITH_EAP
 
 #define TR98_PARM_LAN_DEV_WLAN_CFG "{\
-		\"BeaconType\": { \"Basic\": { }, \"WPA\": { }, \"11i\": { }, \"WPAand11i\": { }, \"Emp\": { } },\
+		\"BeaconType\": { \"Basic\": { }, \"WPA\": { }, \"11i\": { }, \"WPAand11i\": { }, \"WPA3\": { }, \"11iandWPA3\": { }, \"Emp\": { } },\
 		\"DeviceOperationMode\": { \"InfrastructureAccessPoint\": { }, \"Emp\": { } },\
 		\"BasicEncryptionModes\": { \"None\": { }, \"WEPEncryption\": { } },\
 		\"BasicAuthenticationMode\": { \"None\": { }, \"EAPAuthentication\": { }, \"SharedAuthentication\": { } },\
@@ -80,7 +80,7 @@
 #else
 
 #define TR98_PARM_LAN_DEV_WLAN_CFG "{\
-		\"BeaconType\": { \"Basic\": { }, \"WPA\": { }, \"11i\": { }, \"WPAand11i\": { } },\
+		\"BeaconType\": { \"Basic\": { }, \"WPA\": { }, \"11i\": { }, \"WPAand11i\": { }, \"WPA3\": { }, \"11iandWPA3\": { } },\
 		\"DeviceOperationMode\": { \"InfrastructureAccessPoint\": { }, \"Emp\": { } },\
 		\"BasicEncryptionModes\": { \"None\": { }, \"WEPEncryption\": { } },\
 		\"BasicAuthenticationMode\": { \"None\": { } },\
@@ -4866,6 +4866,12 @@
 				if(strstr(json_object_get_string(paramValue), "WPA-WPA2") != NULL) {
 					paramValue = json_object_new_string("WPAand11i");
 				}
+				else if(strstr(json_object_get_string(paramValue), "WPA3-Personal-Transition") != NULL) {
+					paramValue = json_object_new_string("11iandWPA3");
+				}
+				else if(strstr(json_object_get_string(paramValue), "WPA3-Personal") != NULL) {
+					paramValue = json_object_new_string("WPA3");
+				}
 				else if(strstr(json_object_get_string(paramValue), "WPA2") != NULL) {
 					paramValue = json_object_new_string("11i");
 				}
@@ -4993,6 +4999,7 @@
 	char type[12] = {0};
 	struct json_object *tr98ParamValue = NULL;
 	const char *modeEnabled = json_object_get_string(json_object_object_get(apSecJobj, "ModeEnabled"));
+	const char *modesSupported = json_object_get_string(json_object_object_get(apSecJobj, "ModesSupported"));
 	const char *tr98ParmStr_1 = NULL, *tr98ParmStr_2 = NULL;
 	const char *tr181AuthStr = NULL, *tr98Auth11iStr = NULL, *tr98AuthWPAStr = NULL;
 	int modifyAuth = 0, tr181Auth = 0, tr98Auth11i = 0, tr98AuthWPA = 0;
@@ -5096,6 +5103,22 @@
 		if(!strcmp(tr98ParmStr_1, "AESEncryption"))
 			json_object_object_add(apSecJobj, "X_ZYXEL_Encryption", json_object_new_string("aes"));
 	}
+	else if(!strcmp(type, "WPA3")) {
+		if(strstr(modesSupported, "WPA3") !=NULL) {
+			json_object_object_add(apSecJobj, "ModeEnabled", json_object_new_string("WPA3-Personal"));
+		}
+		else {
+			return ZCFG_INVALID_PARAM_VALUE;
+		}
+	}
+	else if(!strcmp(type, "11iandWPA3")) {
+		if(strstr(modesSupported, "WPA3") !=NULL) {
+			json_object_object_add(apSecJobj, "ModeEnabled", json_object_new_string("WPA3-Personal-Transition"));
+		}
+		else {
+			return ZCFG_INVALID_PARAM_VALUE;
+		}
+	}
 
 #ifdef ZYXEL_WIFI_SUPPORT_WEP_WPA_MIXMODE
 	else if(!strcmp(type, "WPAand11i")) {
