Index: libzcfg_fe_tr98-1.0/Makefile
===================================================================
--- libzcfg_fe_tr98-1.0.orig/Makefile	2019-09-12 09:14:19.093557237 +0800
+++ libzcfg_fe_tr98-1.0/Makefile	2019-09-12 09:17:02.284384388 +0800
@@ -104,6 +104,10 @@
 CFLAGS += -DCBT_CUSTOMIZATION
 endif
 
+ifeq ($(CONFIG_ZYXEL_WEB_WIRELESS_NEW_VERSION_ONE),y)
+CFLAGS += -DZYXEL_WEB_WIRELESS_NEW_VERSION_ONE
+endif
+
 ifeq ($(CONFIG_ZYXEL_SUPPORT_TR098_BRCM_BONDING_LINE),y)
 CFLAGS += -DZYXEL_SUPPORT_TR098_BRCM_BONDING_LINE
 endif
Index: libzcfg_fe_tr98-1.0/lanDev.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/lanDev.c	2019-09-12 09:14:19.105563238 +0800
+++ libzcfg_fe_tr98-1.0/lanDev.c	2019-09-12 10:09:25.078500500 +0800
@@ -5565,7 +5565,11 @@
 			apSecJobj = NULL;
 			apSecJobj = zcfgFeJsonMultiObjAppend(RDM_OID_WIFI_ACCESS_POINT_SEC, &apIid, multiJobj, tmpObj);
 			json_object_put(tmpObj);
+#ifdef ZYXEL_WEB_WIRELESS_NEW_VERSION_ONE //US will sync sec when one-ssid enabled by SSID be, so must make passkey applied
+			zcfgFeJsonMultiObjAddParam(multiJobj, RDM_OID_WIFI_ACCESS_POINT_SEC, &apIid, "apply", json_object_new_boolean(true));
+#else
 			zcfgFeJsonMultiObjAddParam(multiJobj, RDM_OID_WIFI_ACCESS_POINT_SEC, &apIid, "apply", json_object_new_boolean(false));
+#endif
 		}
 
 		tmpObj = ssidJobj;
