Index: libzcfg_fe_tr98-1.0/mgmtSrv.c
===================================================================
--- libzcfg_fe_tr98-1.0.orig/mgmtSrv.c	2020-08-19 19:48:09.283649636 +0800
+++ libzcfg_fe_tr98-1.0/mgmtSrv.c	2020-08-19 19:40:09.540024900 +0800
@@ -105,7 +105,7 @@
 			continue;
 		}
 
-		if(!strcmp(paramList->name, "X_ZYXEL_RootQueryWouldNotReplyObjects")){
+		if(!strcmp(paramList->name, "X_ZYXEL_QueryRootButNotRepliedObjects")){
 			const char *objectPathList = json_object_get_string(paramValue);
 			if(!zcfgSetRootQueryNotReplyObjects(objectPathList, 0)){
 				if(!multiJobj)
Index: libzcfg_fe_tr98-1.0/mgmtSrv_parameter.h
===================================================================
--- libzcfg_fe_tr98-1.0.orig/mgmtSrv_parameter.h	2020-08-19 19:48:09.891649618 +0800
+++ libzcfg_fe_tr98-1.0/mgmtSrv_parameter.h	2020-08-19 19:39:28.403024900 +0800
@@ -51,8 +51,8 @@
 	{ "X_ZYXEL_Supplemental_ConnReqPassword", PARAMETER_ATTR_WRITE | PARAMETER_ATTR_PASSWORD, 257, json_type_string},
 	{ "X_ZYXEL_Supplemental_ConnReqPort", PARAMETER_ATTR_WRITE, 0, json_type_uint32},
 #endif
-	{ "X_ZYXEL_ReplyAllObjectsWhileRootQuery", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
-	{ "X_ZYXEL_QueryRootButNotRepliedObjects", PARAMETER_ATTR_WRITE, 513, json_type_string},
+	//{ "X_ZYXEL_ReplyAllObjectsWhileRootQuery", PARAMETER_ATTR_WRITE, 0, json_type_boolean},
+	//{ "X_ZYXEL_QueryRootButNotRepliedObjects", PARAMETER_ATTR_WRITE, 513, json_type_string},
 	//{ "X_ZYXEL_QueryButNotRepliedObjects", PARAMETER_ATTR_WRITE, 513, json_type_string},
 #ifdef ZYXEL_XMPP
     { "SupportedConnReqMethods", PARAMETER_ATTR_READONLY, 32, json_type_string},
