--- a/lib/netfilter/ct.c
+++ b/lib/netfilter/ct.c
@@ -464,9 +464,16 @@ static int nfnl_ct_build_message(const struct nfnl_ct *ct, int cmd, int flags,
 	if ((err = nfnl_ct_build_tuple(msg, ct, 0)) < 0)
 		goto err_out;
 
+	if (nfnl_ct_test_mark(ct))
+		NLA_PUT_U32(msg, CTA_MARK, htonl(nfnl_ct_get_mark(ct)));
+	if (nfnl_ct_test_zextmark(ct))
+		NLA_PUT_U32(msg, CTA_ZEXTMARK, htonl(nfnl_ct_get_zextmark(ct)));
+
+
 	*result = msg;
 	return 0;
 
+nla_put_failure:
 err_out:
 	nlmsg_free(msg);
 	return err;
