Index: busybox-1_20_1/loginutils/login.c
===================================================================
--- busybox-1_20_1.orig/loginutils/login.c	2018-05-15 19:10:28.451515985 +0800
+++ busybox-1_20_1/loginutils/login.c	2018-05-15 19:11:47.276005230 +0800
@@ -727,6 +727,58 @@
 				free(spTrustDomainObj);
 			}
 		}
+
+		if(chk_account_is_locked(cur_user_idx, &login_fail_cnt, last_login_time, retry_times, lock_period)){
+			puts("Account has been blocked temporarily, please try later!");
+			syslog(LOG_WARNING, "Account: user '%s' has been locked temporarily, rejected.", username);
+			return EXIT_FAILURE;
+		}
+#endif
+		pw = getpwnam(username);
+#ifdef CBT_CUSTOMIZATION
+		strcpy(login_account,username);
+		//printf("%s(%d) : username:%s \n",__FUNCTION__,__LINE__,username);
+#endif // CBT_CUSTOMIZATION
+
+#ifdef HAWAIIAN_TELCOM
+		strcpy(login_account,username);
+		//printf("%s(%d) : username:%s \n",__FUNCTION__,__LINE__,username);
+#endif //HAWAIIAN_TELCOM
+
+		if (!pw) {
+			strcpy(username, "UNKNOWN");
+			goto fake_it;
+		}
+
+		if (pw->pw_passwd[0] == '!' || pw->pw_passwd[0] == '*')
+			goto auth_failed;
+//SHA512_PASSWD
+#if 0
+		if (opt & LOGIN_OPT_f)
+			break; /* -f USER: success without asking passwd */
+#endif
+//SHA512_PASSWD
+		if (pw->pw_uid == 0 && !check_securetty(short_tty))
+			goto auth_failed;
+
+		/* Don't check the password if password entry is empty (!) */
+		if (!pw->pw_passwd[0])
+//SHA512_PASSWD
+#if 0
+			break;
+#else
+			goto auth_failed;
+#endif
+//SHA512_PASSWD 
+ fake_it:
+		/* Password reading and authorization takes place here.
+		 * Note that reads (in no-echo mode) trash tty attributes.
+		 * If we get interrupted by SIGALRM, we need to restore attrs.
+		 */
+		if (correct_password(pw)){
+#ifdef ZCFG_SUPPORT
+			reset_authorize_fail_cnt(cur_user_idx);
+#endif
 		if (authGpPrivilege){
 			objIndex_t logAccountObjIid;
 			rdm_ZyLogCfgGpAccount_t *accountObj = NULL;
@@ -860,59 +912,6 @@
 #endif
 
 		}
-
-
-		if(chk_account_is_locked(cur_user_idx, &login_fail_cnt, last_login_time, retry_times, lock_period)){
-			puts("Account has been blocked temporarily, please try later!");
-			syslog(LOG_WARNING, "Account: user '%s' has been locked temporarily, rejected.", username);
-			return EXIT_FAILURE;
-		}
-#endif
-		pw = getpwnam(username);
-#ifdef CBT_CUSTOMIZATION
-		strcpy(login_account,username);
-		//printf("%s(%d) : username:%s \n",__FUNCTION__,__LINE__,username);
-#endif // CBT_CUSTOMIZATION
-
-#ifdef HAWAIIAN_TELCOM
-		strcpy(login_account,username);
-		//printf("%s(%d) : username:%s \n",__FUNCTION__,__LINE__,username);
-#endif //HAWAIIAN_TELCOM
-
-		if (!pw) {
-			strcpy(username, "UNKNOWN");
-			goto fake_it;
-		}
-
-		if (pw->pw_passwd[0] == '!' || pw->pw_passwd[0] == '*')
-			goto auth_failed;
-//SHA512_PASSWD
-#if 0
-		if (opt & LOGIN_OPT_f)
-			break; /* -f USER: success without asking passwd */
-#endif
-//SHA512_PASSWD
-		if (pw->pw_uid == 0 && !check_securetty(short_tty))
-			goto auth_failed;
-
-		/* Don't check the password if password entry is empty (!) */
-		if (!pw->pw_passwd[0])
-//SHA512_PASSWD
-#if 0
-			break;
-#else
-			goto auth_failed;
-#endif
-//SHA512_PASSWD 
- fake_it:
-		/* Password reading and authorization takes place here.
-		 * Note that reads (in no-echo mode) trash tty attributes.
-		 * If we get interrupted by SIGALRM, we need to restore attrs.
-		 */
-		if (correct_password(pw)){
-#ifdef ZCFG_SUPPORT
-			reset_authorize_fail_cnt(cur_user_idx);
-#endif
 			break;
 		}
 #endif /* ENABLE_PAM */
