Index: libzyutil-1.0/zyutil_cofw.c
===================================================================
--- libzyutil-1.0.orig/zyutil_cofw.c	2019-12-24 09:38:50.595599241 +0000
+++ libzyutil-1.0/zyutil_cofw.c	2019-12-24 09:40:53.651598068 +0000
@@ -52,7 +52,11 @@
 			return MODEL_DX5301_B0;
 		}else if (strstr(modelName,"EX5301-B0") != NULL){
 			return MODEL_EX5301_B0;
-		}
+		}else if (strstr(modelName,"DX5301-B1") != NULL){
+                        return MODEL_DX5301_B1;
+                }else if (strstr(modelName,"EX5301-B1") != NULL){
+                        return MODEL_EX5301_B1;
+                }
 	}
 
 	return MODEL_UNDEFINED;
Index: libzyutil-1.0/zyutil_cofw.h
===================================================================
--- libzyutil-1.0.orig/zyutil_cofw.h	2019-12-24 09:38:50.595599241 +0000
+++ libzyutil-1.0/zyutil_cofw.h	2019-12-24 09:41:16.755597847 +0000
@@ -20,7 +20,9 @@
     MODEL_VMG8825_B50B,
     MODEL_AM3100_B0,
     MODEL_DX5301_B0,
-    MODEL_EX5301_B0
+    MODEL_EX5301_B0,
+    MODEL_DX5301_B1,
+    MODEL_EX5301_B1
 } model_name_t;
 
 /*!
