Index: libzyutil-1.0/zyutil_sfp.c
===================================================================
--- libzyutil-1.0.orig/zyutil_sfp.c	2020-04-01 11:24:37.797999234 +0800
+++ libzyutil-1.0/zyutil_sfp.c	2020-04-01 11:26:10.323142615 +0800
@@ -89,6 +89,28 @@
 #define DEBUG_PRINT(fmt, args...)    /* Don't do anything in release builds */
 #endif
 
+#ifdef ZYXEL_I2C_WRITE_API
+int strHex(char *str, unsigned char reg, unsigned char *strBuff)
+{
+    int i,j;
+    int len = strlen(str);
+	unsigned int Hex;
+	char temp[2]= {0};
+
+	//set offset 
+   	memcpy(&strBuff[0], &reg, 1);
+	
+    for( i=0,j=0; i<len/2; i++, j+=2){
+		memset(temp, '\0', 2);
+		sprintf(temp,"%c%c",str[j],str[j+1]);
+		//printf("temp = %s\n",temp);
+		Hex = strtol(temp, NULL, 16);
+		memcpy(&strBuff[i+1], &Hex, 1);
+    }
+	return (i+1);
+}
+#endif
+
 int zyUtil_i2c_open()
 {
     int fd = open(DEFAULT_I2C_BUS, O_RDWR);
@@ -130,6 +152,38 @@
     return write(fd, buf, bufsize);
 }
 
+#ifdef ZYXEL_I2C_WRITE_API
+int zyUtil_i2c_write(int fd, unsigned char reg, char *buf){
+	int ret = -1;
+	char buffer[MAX_READ_BYTES];
+	int hexLength, i;
+	
+	int msgLength = strlen(buf);
+	if (msgLength > MAX_READ_BYTES){
+		printf("*** zyUtil_i2c_write Error: Exceeded supported write length = %d \n", MAX_READ_BYTES);
+		return ret;
+	}
+
+	printf("*** zyUtil_i2c_write : wrote '%s' to 0x%02x \n", buf , reg);
+	hexLength = strHex(buf, reg, (unsigned char*)buffer);
+
+	//printf("\n*** zyUtil_i2c_write buffer ***\n", hexLength);
+	printf("Hex: ");
+	for(i=0; i<hexLength; i++){
+		printf("%02x ", (unsigned char)buffer[i]);
+	}
+	printf("\n");
+	//printf("\n*** zyUtil_i2c_write : %d read buffer end *****\n", hexLength);
+
+	ret = _zyUtil_i2c_write_bus(fd, buffer, hexLength);
+	if (ret < 0) {
+		printf("*** Failed to write reg: %02x.\n", buffer);
+		return ret;
+	}
+	printf("*** zyUtil_i2c_write : wrote %d bytes\n\n", ret);
+	return ret;
+}
+#endif
 int zyUtil_i2c_read_register(int fd, unsigned char reg, unsigned char *buf, int bufsize)
 {
     int ret;
Index: libzyutil-1.0/zyutil_sfp.h
===================================================================
--- libzyutil-1.0.orig/zyutil_sfp.h	2020-04-01 11:24:37.661999229 +0800
+++ libzyutil-1.0/zyutil_sfp.h	2020-04-01 11:26:19.183544236 +0800
@@ -67,5 +67,9 @@
  *
  */
 int zyUtil_sfp_read_ddmi_state(sfp_ddmi_state_t *ddmi_state);
+#ifdef ZYXEL_I2C_WRITE_API
+int zyUtil_i2c_write(int fd, unsigned char reg, char *buf);
+int strHex(char *str, unsigned char reg, unsigned char *strBuff);
+#endif
 
 #endif
