Index: libzyutil-1.0/zyutil_sfp.c
===================================================================
--- libzyutil-1.0.orig/zyutil_sfp.c	2020-04-08 10:45:37.689576576 +0800
+++ libzyutil-1.0/zyutil_sfp.c	2020-04-08 10:48:03.792919473 +0800
@@ -289,19 +289,32 @@
 
 int dump_DDMI_state(sfp_ddmi_state_t *ddmi_state)
 {
+	
+#ifdef WIND_ITALY_CUSTOMIZATION
+	unsigned char gpon_sn[DDMI_STRING_MAX_LENGTH+1];
+
+	memset(gpon_sn, '\0', sizeof(gpon_sn) );
+	strncpy(gpon_sn, ddmi_state->sfp_sn ,DDMI_STRING_MAX_LENGTH);
+#else
+	dumpString( &ddmi_state->sfp_sn, DDMI_STRING_MAX_LENGTH);
+#endif
 	dumpString( &ddmi_state->sfp_vendor, DDMI_STRING_MAX_LENGTH);
 	dumpString( &ddmi_state->sfp_pn, DDMI_STRING_MAX_LENGTH);
-	dumpString( &ddmi_state->sfp_sn, DDMI_STRING_MAX_LENGTH);
 	
 	printf("DDMI state:\n");
 	printf("sfp presetnt      = %d\n",        ddmi_state->present);
 	printf("sfp vendor name   = %s\n",        ddmi_state->sfp_vendor);
 	printf("sfp part number   = %s\n",        ddmi_state->sfp_pn);
+#ifdef WIND_ITALY_CUSTOMIZATION
+	printf("sfp serial number = %s\n",        gpon_sn);
+#else
 	printf("sfp serial number = %s\n",        ddmi_state->sfp_sn);
+#endif
 	printf("sfp rx power      = %f dbm\n",    ddmi_state->rx_power);
 	printf("sfp tx power      = %f dbm\n",    ddmi_state->tx_power);
 	printf("sfp temperature   = %f C\n",      ddmi_state->temperature);
-    return 0;
+	
+	return 0;
 }
 
 
