Index: libzyutil-1.0/ECONET/libzyutil_wrapper.c
===================================================================
--- libzyutil-1.0.orig/ECONET/libzyutil_wrapper.c	2020-01-17 03:37:14.009676952 +0000
+++ libzyutil-1.0/ECONET/libzyutil_wrapper.c	2020-01-17 03:39:14.673675802 +0000
@@ -166,6 +166,10 @@
 	char username[16] = {0};
 	void *sysCmd_reply=NULL;
 	char filename[32] = {0};
+#if 1 //IChia 20200117
+	int  fd_template;
+	FILE *fp_template;
+#endif
 
 	// clean the MRD info
 	memset(mrd, 0x0, sizeof(MRD_DATA));
@@ -175,7 +179,11 @@
 	cuserid(username); 
 
 	strcpy(filename, MRDINFO_FILE_X);
+#if 1 //IChia 20200117
+	if ((fd_template = mkstemp(filename)) < 0 ){
+#else
 	if (mkstemp(filename) < 0 ){
+#endif
 		printf("%s: Can't create MRD file!\n",__FUNCTION__);
 		return ZCFG_INTERNAL_ERROR;
 	} 
@@ -201,11 +209,16 @@
     }
 
     /*Read MRD info to buffer*/
+#if 1  //IChia 20200117
+    if ((fp_template = fdopen(fd_template, "rb")) == NULL) {
+#else
     if((fptr = fopen(filename, "rb")) == NULL) {
+#endif
         printf("%s: Can't fopen %s\n", __FUNCTION__,filename);
 		unlink(filename);
         return ZCFG_INTERNAL_ERROR;
     }
+
 #else
 	/* check MRDINFO_FILE exists or not. */
 	if ((fptr = fopen(MRDINFO_FILE, "rb")) == NULL)
@@ -218,7 +231,12 @@
 	}
 	/*Read MRD info to buffer*/
 #endif
+
+#if 1 //IChia 20200117
+	fread(mrd, 1, (uint32_t)MRDINFO_SIZE, fp_template);
+#else
 	fread(mrd, 1, (uint32_t)MRDINFO_SIZE, fptr);
+#endif
 #ifdef ZYXEL_DETECT_PRODUCT_TYPE
 	char prname[MRD_PRODUCT_NAME_LEN] = {0};
 	int prnameLen = MRD_PRODUCT_NAME_LEN;
@@ -245,7 +263,11 @@
 #endif
 
 	dumpMRDinfo(mrd);
+#if 1 //IChia 20200117
+	fclose(fp_template);
+#else
 	fclose(fptr);
+#endif
 	unlink(filename);
 	return ZCFG_SUCCESS;
 }
