Index: pure-ftpd-1.0.30/src/ftpd.c
===================================================================
--- pure-ftpd-1.0.30.orig/src/ftpd.c	2020-04-24 15:28:12.248465890 +0800
+++ pure-ftpd-1.0.30/src/ftpd.c	2020-04-24 15:29:07.891982781 +0800
@@ -1940,6 +1940,71 @@
 }
 #endif //ZYXEL_REMOMGMT_ONLY_HTTPS_ALLOWS_FOR_WLAN_CONNECT
 
+#ifdef ZYXEL_REMOTE_ACCESS_PRIVILEGE
+void checkClientFromLANWAN(char *remoaddr, char *remoAddr_Type)
+{
+	int addrType = 0;
+	objIndex_t objIid;
+	objIndex_t v4AddrIid;
+	objIndex_t v6AddrIid;
+	rdm_IpIface_t *ipObj = NULL;
+	rdm_IpIfaceV4Addr_t *ipv4Obj = NULL;
+	rdm_IpIfaceV6Addr_t *ipv6Obj = NULL;
+	char lanip[64] = {0};
+	char perfix[8] = {0};
+	bool found = false;
+	
+	addrType = ipType(remoaddr);
+	if (addrType != AF_INET && addrType != AF_INET6) {
+		strcpy(remoAddr_Type, "Unknown");
+		logfile(LOG_WARNING, "[%s][%d] Unknow IP type!!\n", __FUNCTION__, __LINE__);
+		return;
+	}
+	
+	/*LANIP*/
+	IID_INIT(objIid);
+	while(!found && zcfgFeObjStructGetNext(RDM_OID_IP_IFACE, &objIid, (void **)&ipObj) == ZCFG_SUCCESS) {
+		if(!strcmp(ipObj->Name, "LAN")) {
+			if(addrType == AF_INET) {
+				IID_INIT(v4AddrIid);
+				while(zcfgFeObjStructGetNext(RDM_OID_IP_IFACE_V4_ADDR, &v4AddrIid, (void **)&ipv4Obj) == ZCFG_SUCCESS) {
+					if(objIid.idx[0] == v4AddrIid.idx[0]) { //check LAN group	
+						if (!zyUtilIp4AddrDomainCmp(ipv4Obj->IPAddress, ipv4Obj->SubnetMask, remoaddr)) {
+							found = true;
+							zcfgFeObjStructFree(ipv4Obj);
+							break;
+						}
+					}
+					zcfgFeObjStructFree(ipv4Obj);
+				}
+			}
+			else {
+				IID_INIT(v6AddrIid);
+				while(zcfgFeObjStructGetNext(RDM_OID_IP_IFACE_V6_ADDR, &v6AddrIid, (void **)&ipv6Obj) == ZCFG_SUCCESS) {
+					if(objIid.idx[0] == v6AddrIid.idx[0]) { //check LAN group
+						sscanf(ipv6Obj->IPAddress, "%[^/]/%s", lanip, perfix);
+						if (checkCidrBlock(lanip, perfix, remoaddr) == 1) {
+							found = true;
+							zcfgFeObjStructFree(ipv6Obj);
+							break;
+						}
+					}
+					zcfgFeObjStructFree(ipv6Obj);
+				}
+			}
+		}
+		zcfgFeObjStructFree(ipObj);
+	}
+	
+	if (found) {
+		strcpy(remoAddr_Type, "LAN");
+	}
+	else {
+		strcpy(remoAddr_Type, "WAN");
+	}
+}
+#endif
+
 void dopass(char *password)
 {
     static unsigned int tapping;    
@@ -2061,88 +2126,21 @@
 #endif //ZYXEL_REMOMGMT_ONLY_HTTPS_ALLOWS_FOR_WLAN_CONNECT
 
 #ifdef ZYXEL_REMOTE_ACCESS_PRIVILEGE
-
-		//logfile(LOG_WARNING, "Account privilege: ZYXEL_REMOTE_ACCESS_PRIVILEGE IN !!!!");
-		//logfile(LOG_WARNING, "Account privilege: get remoAddr = %s ", hostip);
-
-		// check remote IP v4/v6
-		int addrType = 0;
-		char remoAddr_Type[32] = {0}; //WAN LAN
-		addrType = ipType(hostip);
-		
-		if (addrType == AF_INET){
-			//v4 checking
-			
-			//logfile(LOG_WARNING, "Account privilege: %s remoAddr type = %d ", hostip, addrType);
-			
-			/*Remote Ip WAN/LAN Ip check */
-			
-			//Get host IP	
-			int n;
-			int remoAddrType = 0;
-			objIndex_t objIid;
-			rdm_IpIface_t *ipObj = NULL;
-			rdm_IpIfaceV4Addr_t *ipv4Obj = NULL;	
-			char lanip[32] = {0};
-			char subnetMask[32] = {0};
-			char perfix[10]={0};
-
-			/*LANIP*/
-			IID_INIT(objIid);
-			while( zcfgFeObjStructGetNext(RDM_OID_IP_IFACE, &objIid, (void **)&ipObj) == ZCFG_SUCCESS) {
-				if(strcmp(ipObj->X_ZYXEL_IfName, "br0") ==0) {
-					free(ipObj);
-					break;
-				}
-				free(ipObj);
-			}
-			
-			bool checkPrivilege = false;
-			objIndex_t v4AddrIid;
-			v4AddrIid.level = 2;
-			
-			IID_INIT(v4AddrIid);
-			while( zcfgFeObjStructGetNext(RDM_OID_IP_IFACE_V4_ADDR, &v4AddrIid, (void **)&ipv4Obj)== ZCFG_SUCCESS){
-				if (objIid.idx[0] == v4AddrIid.idx[0]){ //check br0(LAN) group
-					strcpy(lanip, ipv4Obj->IPAddress);
-					strcpy(subnetMask, ipv4Obj->SubnetMask);
-					
-					inet_pton(AF_INET, subnetMask, &n);
-					int i = 0;
-
-					while (n > 0) {
-							n = n >> 1;
-							i++;
-					}
-					sprintf(perfix, "%d", i);
-					if (checkCidrBlock(lanip, perfix, hostip) == 1){
-						strcpy(remoAddr_Type, "LAN");
-					}
-					else{
-						strcpy(remoAddr_Type, "WAN");
-					}
-					
-					if (strstr(userPriviege, remoAddr_Type) != NULL){
-						checkPrivilege = true;
-					}
-				}
-				free(ipv4Obj);
+		bool checkPrivilege = false;
+		char remoAddr_Type[32] = {0};
+		if(strcmp(userPriviege, "LAN,WAN")) {
+			checkClientFromLANWAN(hostip, remoAddr_Type);
+			if (strstr(userPriviege, remoAddr_Type) != NULL) {
+				checkPrivilege = true;
 			}
 
-			if (checkPrivilege == false){
+			if (checkPrivilege == false) {
 				snprintf(logStr, sizeof(logStr), "Account:'%s' FTP permission denied.", account);
 				puts(logStr);
 				logfile(LOG_WARNING, "Account privilege:'%s' FTP permission denied.", account);
 				return;
-			}		
-		}
-		else if (addrType == AF_INET6){
-			//not support V6 for WIND Italy
-		}
-		else{
-			//skip check unknow IP type
+			}
 		}
-
 #endif
 	}
 
