Index: usb-modeswitch-2.1.0/Makefile
===================================================================
--- usb-modeswitch-2.1.0.orig/Makefile	2014-01-28 07:06:11.000000000 +0800
+++ usb-modeswitch-2.1.0/Makefile	2019-09-26 18:26:42.000293397 +0800
@@ -1,7 +1,7 @@
 PROG        = usb_modeswitch
 VERS        = 2.1.0
 CC          ?= gcc
-CFLAGS      += -Wall
+CFLAGS      += -Wall -idirafter$(STAGING_DIR)/usr/include/libusb-1.0
 LIBS        = `pkg-config --libs --cflags libusb-1.0`
 RM          = /bin/rm -f
 OBJS        = usb_modeswitch.c
