Index: zoneinfo-2018d/etcetera
===================================================================
--- zoneinfo-2018d.orig/etcetera	2016-08-24 07:52:15.000000000 +0800
+++ zoneinfo-2018d/etcetera	2019-08-16 15:14:20.085728731 +0800
@@ -55,18 +55,21 @@
 Zone	Etc/GMT-12	12	-	+12
 Zone	Etc/GMT-11	11	-	+11
 Zone	Etc/GMT-10	10	-	+10
+Zone	Etc/GMT-930	9:30	-	+09:30
 Zone	Etc/GMT-9	9	-	+09
 Zone	Etc/GMT-8	8	-	+08
 Zone	Etc/GMT-7	7	-	+07
 Zone	Etc/GMT-6	6	-	+06
 Zone	Etc/GMT-5	5	-	+05
 Zone	Etc/GMT-4	4	-	+04
+Zone	Etc/GMT-330	3:30	-	+03:30
 Zone	Etc/GMT-3	3	-	+03
 Zone	Etc/GMT-2	2	-	+02
 Zone	Etc/GMT-1	1	-	+01
 Zone	Etc/GMT+1	-1	-	-01
 Zone	Etc/GMT+2	-2	-	-02
 Zone	Etc/GMT+3	-3	-	-03
+Zone	Etc/GMT+330	-03:30	-	-03:30
 Zone	Etc/GMT+4	-4	-	-04
 Zone	Etc/GMT+5	-5	-	-05
 Zone	Etc/GMT+6	-6	-	-06
