Index: ntpclient-2007/Makefile
===================================================================
--- ntpclient-2007.orig/Makefile	2019-08-28 13:52:54.805994479 +0800
+++ ntpclient-2007/Makefile	2019-08-28 13:54:03.556648928 +0800
@@ -15,7 +15,7 @@
 # CFLAGS += -DUSE_OBSOLETE_GETTIMEOFDAY
 CFLAGS += -DZyXEL_PATCH
 
-ZYXEL_LIBS = -lzcfg_msg
+ZYXEL_LIBS = -lzcfg_msg -lzlog -lzos
 LDFLAGS += -lrt $(ZYXEL_LIBS)
 
 all: ntpclient
